include/rocksdb/fs_env.h

```diff
@@ -0,0 +1,1323 @@
+// An FSEnv is an interface used by the rocksdb implementation to access
+// storage functionality like the filesystem etc.  Callers
+// may wish to provide a custom FSEnv object when opening a database to
+// get fine gain control; e.g., to rate limit file system operations.
+//
+// All FSEnv implementations are safe for concurrent access from
+// multiple threads without any external synchronization.
+
+#pragma once
+
+#include <stdint.h>
+#include <cstdarg>
+#include <functional>
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+#include "rocksdb/env.h"
+#include "rocksdb/io_status.h"
+#include "rocksdb/thread_status.h"
+
+namespace rocksdb {
+
+class FileLock;
+class FSDirectory;
+class FSMemoryMappedFileBuffer;
+class FSRandomAccessFile;
+class FSRandomRWFile;
+class FSSequentialFile;
+class FSWritableFile;
+class Logger;
+class Slice;
+struct DBOptions;
+struct ImmutableDBOptions;
+struct MutableDBOptions;
+class RateLimiter;
+
+// The FSEnv, FSSequentialFile, FSRandomAccessFile, FSWritableFile,
+// FSRandomRWFileclass, and FSDIrectory classes define the interface between
+// RocksDB and storage systems, such as Posix filesystems,
+// remote filesystems etc.
+// The interface allows for fine grained control of individual IO operations,
+// such as setting a timeout, prioritization, hints on data placement,
+// different handling based on type of IO etc.
+// This is accomplished by passing an instance of IOOptions to every
+// API call that can potentially perform IO. Additionally, each such API is
+// passed a pointer to a IODebugContext structure that can be used by the
+// storage system to include troubleshooting information. The return values
+// of the APIs is of type IOStatus, which can indicate an error code/sub-code,
+// as well as metadata about the error such as its scope and whether its
+// retryable.
+class FSEnv {
+ public:
+  struct FileAttributes {
+    // File name
+    std::string name;
+
+    // Size of file in bytes
+    uint64_t size_bytes;
+  };
+
+  // Priority of an IO request. This is a hint and does not guarantee any
+  // particular QoS.
+  // IO_LOW - Typically background reads/writes such as compaction/flush
+  // IO_HIGH - Typically user reads/synchronous WAL writes
+  enum class IOPriority : uint8_t {
+    IO_LOW,
+    IO_HIGH,
+    IO_TOTAL,
+  };
+
+  // Type of the data begin read/written. It can be passed down as a flag
+  // for the FSEnv implementation to optionally handle different types in
+  // different ways
+  enum class IOType : uint8_t {
+    kData,
+    kFilter,
+    kIndex,
+    kMetadata,
+    kWAL,
+    kManifest,
+    kLog,
+    kUnknown,
+    kInvalid,
+  };
+
+  // Per-request options that can be passed down to the FSEnv implementation.
+  // These are hints and are not necessarily guaranteed to be honored.
+  // More hints can be added here in the future to indicate things like
+  // storage media (HDD/SSD) to be used, replication level etc.
+  struct IOOptions {
```

Comment:
I can see where you might want this on a per-file basis, but why do you want them on a per-request basis?  Can you give an example where one read/write from a file would be a different priority than another to the same file?

Reply:
The timeout might be a per-operation parameter, but the IOType is not likely to be (are you suggesting the same file might be used for different types of IO?  I also doubt that read or write operations to a single IO file would have different priorities (this read his high but the next one is low does not make sense to me).

Reply:
We use a single TableReader object for user reads and compaction reads, so yes its possible to have different priority IOs to the same file. Even if we use a separate TableReader for compaction, I wonder if we might want to allow users to specify a priority. All user reads would go through the same file.

Reply:
Why use IOOptions for this though, rather than using the ReadOptions that are already present?  Why add a new struct with new parameters rather than extending the one that already exists?  It seems like there are an awful lot of places where you would need to update (SST reader/writer, trace methods, etc)